File: code/python/剑指offer18.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None
class Solution:
    def deleteNode(self, head: ListNode, val: int) -> ListNode:
        if head.next is None:
            return None
        if head.val==val:
            return head.next
        last = head
        cur = head.next
        while cur.val!=val:
            last = cur
            cur = cur.next
        last.next = cur.next
        return head

File: code/java/剑指offer18.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
class Solution {
    public ListNode deleteNode(ListNode head, int val) {
        if(head.val == val) return head.next;
        ListNode pre = head;
        ListNode node = head;
        while(node!=null){
            if(node.val==val) break;
            pre = node;
            node = node.next;
        }
        pre.next = node.next;
        node.next = null;
        return head;
    }
}